### PR TITLE
Add group support

### PR DESCRIPTION
**Pull request by alexhung**
As part of #1:

- Add ability to CRUD group for project.
- Generalized `user` resource into `membership` as `group` shares identical HCL/JSON structure, except the API endpoints.
- Use [terraform-plugin-docs](https://github.com/hashicorp/terraform-plugin-docs) to generate documentation. It generates a slightly different format than the one TF recommends (https://www.terraform.io/docs/registry/providers/docs.html#format) but this plugin is the [recommended one to use](https://www.terraform.io/docs/registry/providers/docs.html#generating-documentation). 🤷

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -30,90 +30,73 @@ The following 3 license types (`jq .type`) do **NOT** support APIs:
 - OSS
 
 ## Example Usage
-```hcl
+
+```terraform
```
Comment (chb0github) [inline docs/index.md:34]: hcl format isn't a thing?

File: docs/index.md [hunk 2]
```diff
@@ -30,90 +30,73 @@ The following 3 license types (`jq .type`) do **NOT** support APIs:
...
+    name  = "user2"
+    roles = ["developer"]
+  }
 
-### Basic Auth
-Basic auth may be used by adding a `username` and `password` field to the provider block
-Getting this value from the environment is supported with the `ARTIFACTORY_USERNAME` and `ARITFACTORY_PASSWORD` variable
+  group {
+    name = "dev-group"
+    roles = ["developer"]
```
Comment (chb0github) [inline docs/index.md:70]: are you able to relate a member to a group? That'd be how to demonstrate

File: docs/index.md (hunk 1 above)
Comment (alexhung) [reply docs/index.md:34]: I think both are supported. The plugin uses `terraform` 🤷‍♂️

File: docs/resources/project.md [hunk 3]
```diff
@@ -1,45 +1,97 @@
...
+resource "project" "myproject" {
+  key = "myproj"
+  display_name = "My Project"
+  description  = "My Project"
   admin_privileges {
     manage_members   = true
     manage_resources = true
     index_resources  = true
   }
+  max_storage_in_gibabytes   = 10
```
Comment (chb0github) [inline docs/resources/project.md:25]: misspelled gigabytes

File: docs/resources/project.md [hunk 4]
```diff
@@ -1,45 +1,97 @@
...
+
+  group {
+    name = "release-group"
+    roles = ["release manager"]
+  }
 }
 ```
 
-## Argument Reference
+<!-- schema generated by tfplugindocs -->
```
Comment (chb0github) [inline docs/resources/project.md:51]: yeah!

File: docs/index.md (hunk 2 above)
Comment (alexhung) [reply docs/index.md:70]: Group's membership is not managed by this provider. Customers will have to managed the group's composition via either the UI/REST/TF RT provider separately.

File: docs/resources/project.md [hunk 5]
```diff
@@ -1,45 +1,97 @@
...
+
+### Required
+
+- **admin_privileges** (Block Set, Min: 1) (see [below for nested schema](#nestedblock--admin_privileges))
+- **display_name** (String) Also known as project name on the UI
+- **key** (String) The Project Key is added as a prefix to resources created within a Project. This field is mandatory and supports only 3 - 6 lowercase alphanumeric characters. Must begin with a letter. For example: us1a.
+
+### Optional
+
+- **block_deployments_on_limit** (Boolean)
```
Comment (chb0github) [inline docs/resources/project.md:62]: description here

File: docs/resources/project.md [hunk 6]
```diff
@@ -1,45 +1,97 @@
...
+- **description** (String)
+- **email_notification** (Boolean) Alerts will be sent when reaching 75% and 95% of the storage quota. Serves as a notification only and is not a blocker
+- **group** (Block Set) Project group. Element has one to one mapping with the [JFrog Project Groups API](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-UpdateGroupinProject) (see [below for nested schema](#nestedblock--group))
+- **id** (String) The ID of this resource.
+- **max_storage_in_gibabytes** (Number) Storage quota in GB. Must be 1 or larger. Set to -1 for unlimited storage.
```
Comment (chb0github) [inline docs/resources/project.md:67]: misspelled

File: docs/resources/project.md (hunk 5 above)
Comment (alexhung) [reply docs/resources/project.md:62]: I will add to the schema. Nothing from UI so I'll have to be creative 😄

File: pkg/projects/resource_project_membership.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type Equatable interface {
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:11]: I think you should implement `Identifiable` on all of them

File: pkg/projects/resource_project_membership.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+// Use by both project user and project group, as they shared identical data structure
+type Membership struct {
+	Members []Member
+}
+
+func (a Member) Equals(b Member) bool {
+	return a.Name == b.Name
+}
+
+func contains(as []Member, b Member) bool {
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:30]: You can actually define this on `Membership`

File: pkg/projects/resource_project_membership.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	return members
+}
+
+var unpackMembers = func(data *schema.ResourceData, membershipKey string) (string, Membership, error) {
+	d := &ResourceData{data}
+	projectKey := d.getString("key", false)
+
+    membership := Membership{
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:95]: formatting. Run the go formatter

File: pkg/projects/resource_project_membership.go [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+	membersToBeAdded := membershipDifference(terraformMembership.Members, projectMembers)
+	log.Printf("[TRACE] membersToBeAdded: %+v\n", membersToBeAdded)
+	membersToBeUpdated := membershipIntersection(terraformMembership.Members, projectMembers)
+	log.Printf("[TRACE] membersToBeUpdated: %+v\n", membersToBeUpdated)
+	membersToBeDeleted := membershipDifference(projectMembers, terraformMembership.Members)
+	log.Printf("[TRACE] membersToBeDeleted: %+v\n", membersToBeDeleted)
+
+	var errs []error
+
+	for _, member := range append(membersToBeAdded, membersToBeUpdated...) {
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:158]: nice

File: pkg/projects/resource_project_membership.go [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+	var errs []error
+	for _, member := range members {
+		err := deleteMember(membershipUrl, member, m)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if errs != nil && len(errs) > 0 {
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:197]: I don't think you need to check for nil. I think `len` does that for you

File: pkg/projects/resource_project_membership.go (hunk 8 above)
Comment (alexhung) [reply pkg/projects/resource_project_membership.go:30]: Oh yeah. Let me try that.

File: pkg/projects/resource_project_membership.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+		return fmt.Errorf("failed to delete members from project: %s", errs)
```
Comment (chb0github) [inline pkg/projects/resource_project_membership.go:198]: There is a format code for using arrays - I think it's `%q` - check rt code

File: pkg/projects/resource_project_membership.go (hunk 7 above)
Comment (alexhung) [reply pkg/projects/resource_project_membership.go:11]: `Equatable` is what you call `Identifiable` :)

File: pkg/projects/resource_project.go [hunk 13]
```diff
@@ -217,19 +252,24 @@ func projectResource() *schema.Resource {
...
+
+		return packProject(data, &project, users, groups)
 	}
 
 	var createProject = func(ctx context.Context, data *schema.ResourceData, m interface{}) diag.Diagnostics {
 		log.Printf("[DEBUG] createProject")
 		log.Printf("[TRACE] %+v\n", data)
 
-		key, project, users, err := unpackProject(data)
+		key, project, users, groups, err := unpackProject(data)
```
[the remark this answers is not in the document]
Comment (alexhung) [reply pkg/projects/resource_project.go:272]: Good idea.

File: tools/tools.go [hunk 14]
```diff
@@ -0,0 +1,8 @@
+// +build tools
```
Comment (chb0github) [inline tools/tools.go:1]: What's this file about?

**Review by chb0github — COMMENTED** [review]
misc questions

File: tools/tools.go (hunk 14 above)
Comment (alexhung) [reply tools/tools.go:1]: This is how utility package is defined so `go` cli can use them

Comment (alexhung) [reply tools/tools.go:1]: https://github.com/go-modules-by-example/index/blob/master/010_tools/README.md

**Review by chb0github — APPROVED** [review]

**Comment by alexhung** [comment]
Per conversation with @chb0github, I'll improve the description for `max_storage_in_gibibytes` in next PR and expand on the 2^x based unit.

**Pull request merged**